text/0000-clustering.md

```diff
@@ -0,0 +1,218 @@
+- RFC PR: [tremor-rs/tremor-rfcs#0000](https://github.com/tremor-rs/tremor-rfcs/pull/0000)
+
+# Summary
+[summary]: #summary
+
+TODO expand
+
+Implement basic clustering for Tremor.
+
+# Motivation
+[motivation]: #motivation
+
+TODO expand more
+
+Add clustering to Tremor, so that it’s a true distributed system. This will add
+resiliency for the pipelines deployed under Tremor, in terms of both host
+failures as well as application logic (eg: correct distributed throttling for
+logs/metrics even when some Tremor nodes go down). Moreover, it allows for
+denser Tremor deployments with regards to the total host footprint (thus saving
+more on hardware resources), and also eases the propagation/synchronization of
+changes that need to touch all hosts, opening up nicer ways of doing tremor
+pipeline deployments.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Tremor can be operated on a clustered mode, which allows operators to use a
+group of tremor nodes collaboratively for their event processing needs, giving
+the illusion of a single running tremor instance to the outside world.
+
+A tremor cluster is composed of two sets of nodes: `coordinators` and `workers`.
+
+Coordinator nodes are the ones coordinating the addition and removal of nodes to
+the cluster. They also intercept all changes to a running cluster (basically new
+workload submissions, or changes to the tremor [repository and registry
+state](https://docs.tremor.rs/operations/configuration/#introduction)) and
+ensure that all nodes eventually apply the changes -- anytime the [tremor
+api](https://docs.tremor.rs/api/) used to submit changes from a node, the
+request is routed first to the coordinator node (if the node accepting the
+request is not already a coordinator). Thus, the group of coordinator nodes can
+be thought of as the brain of the cluster. For the initial clustering
+implementation, the allowed number of dedicated coordinator nodes is 3.
+
+If coordinators form the brain, worker nodes are the brawn of the cluster,
+hosting the actual running instances of the deployed sources, pipelines and
+sinks. Worker nodes accept changes from the coordinator nodes for this and
+communicate directly to the cluster-external world as their workload demands
+(i.e. as part of the sources/sinks running on the node). There is no set limit
+on the number of worker nodes in the cluster, though it will be bound in
+practicality by the network constraints of cluster-wide communication.
+
+An example of starting a tremor server process in these two modes:
+
+```sh
+# assuming this command is run on host1 and the 3 coordinator nodes are on
+# host1, host2, host3 (host1 can be left out here if desired)
+tremor server run --coordinator --peers "host1:8139,host2:8139,host3:8139"
+
+# assuming this command is run on host4 (all 3 coordinator nodes have to be specified)
+tremor server run --worker --coordinators "host1:8139,host2:8139,host3:8139"
```

Comment:
For simplicity, and for small clusters, mostly for testing, wouldn't it make sense to allow a node to be coordinator and worker at the same time. This way we can form a simple cluster with 3 nodes where either 1 or all three are coordinators.

Reply:
I agree, it would, that would also remove a extra mainline for a non clustered version (it'd effectively be a cluster of 1)